ns for fx: allow graph matching of parents of cat

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52799 ns for fx: remove ".stats" suffix
* #52798 ns for fx: add node name and type to results
* #52789 ns for fx: make return type of ns APIs future proof
* #52779 ns for fx: unify return types of weight and activation APIs
* #52771 ns for fx: decouple subgraph names from node names
* #52681 ns for fx: update graph matching to handle dicts and tuples in node args
* #52402 ns for fx: update graph matching to not match nodes with equal types
* #52395 ns for fx: support linear_relu for weight matching
* **#52368 ns for fx: allow graph matching of parents of cat**

Summary:

Before this PR, the graph matching logic only handles node arguments of
type Node. This PR extends it to allow to handle node arguments of type
Tuple, so that the matcher can properly navigate through the arguments
of `cat`.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_before_cat
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26490101](https://our.internmc.facebook.com/intern/diff/D26490101)

vkuzo committed:
ns for fx: allow graph matching of parents of cat

Summary:

Before this PR, the graph matching logic only handles node arguments of
type Node. This PR extends it to allow to handle node arguments of type
Tuple, so that the matcher can properly navigate through the arguments
of `cat`.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_before_cat
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -573,6 +573,34 @@ def test_matching_failure_node_type(self):
+        results = get_matching_subgraph_pairs(mp, mq)
+
+        expected_types = {
+            'cat_1': ((torch.cat, torch.cat), (toq.cat, toq.cat)),
```
Why are there 2 cats here when the graph has a single cat node?

vkuzo replied:
tl;dr;
* we need to use subgraphs to properly support fusions
* the current definition of subgraph is (start_node, end_node) (this may change in the future for more complicated subgraphs)
* this code is for a subgraph of one node, with start_node being the same as end_node

vkuzo committed:
rebase on "ns for fx: allow graph matching of parents of cat"


Summary:

Before this PR, the graph matching logic only handles node arguments of
type Node. This PR extends it to allow to handle node arguments of type
Tuple, so that the matcher can properly navigate through the arguments
of `cat`.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_before_cat
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26490101](https://our.internmc.facebook.com/intern/diff/D26490101)

[ghstack-poisoned]

vkuzo committed:
fix asan on "ns for fx: allow graph matching of parents of cat"


Summary:

Before this PR, the graph matching logic only handles node arguments of
type Node. This PR extends it to allow to handle node arguments of type
Tuple, so that the matcher can properly navigate through the arguments
of `cat`.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_before_cat
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26490101](https://our.internmc.facebook.com/intern/diff/D26490101)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: allow graph matching of parents of cat"


Summary:

Before this PR, the graph matching logic only handles node arguments of
type Node. This PR extends it to allow to handle node arguments of type
Tuple, so that the matcher can properly navigate through the arguments
of `cat`.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_before_cat
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26490101](https://our.internmc.facebook.com/intern/diff/D26490101)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: allow graph matching of parents of cat"


Summary:

Before this PR, the graph matching logic only handles node arguments of
type Node. This PR extends it to allow to handle node arguments of type
Tuple, so that the matcher can properly navigate through the arguments
of `cat`.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_before_cat
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26490101](https://our.internmc.facebook.com/intern/diff/D26490101)

[ghstack-poisoned]
